indra/newview/gltf/primitive.cpp

```diff
@@ -30,9 +30,204 @@
 #include "buffer_util.h"
 #include "../llviewershadermgr.h"
 
+#include "mikktspace/mikktspace.hh"
+
+#include "meshoptimizer/meshoptimizer.h"
+
+
 using namespace LL::GLTF;
 using namespace boost::json;
 
+
+// Mesh data useful for Mikktspace tangent generation (and flat normal generation)
+struct MikktMesh
+{
+    std::vector<LLVector3> p;
+    std::vector<LLVector3> n;
+    std::vector<LLVector2> tc;
+    std::vector<LLVector4> w;
+    std::vector<LLVector4> t;
+    std::vector<LLColor4U> c;
+    std::vector<U64> j;
+
+    // initialize from src primitive and make an unrolled triangle list
+    // returns false if the Primitive cannot be converted to a triangle list
+    bool copy(const Primitive* prim)
+    {
+        bool indexed = !prim->mIndexArray.empty();
+        U32 vert_count = indexed ? prim->mIndexArray.size() : prim->mPositions.size();
+
+        if (prim->mMode != Primitive::Mode::TRIANGLES)
+        {
+            LL_WARNS("GLTF") << "Unsupported primitive mode for conversion to triangles: " << (S32) prim->mMode << LL_ENDL;
+            return false;
+        }
+
+        p.resize(vert_count);
+        n.resize(vert_count);
+        tc.resize(vert_count);
+        c.resize(vert_count);
```

Comment:
Resizes called unnecessarily here for vertex attributes which code below suggests may not exist.